src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -346,11 +343,35 @@ private bool IsLegacyAdalCacheEnabled(AuthenticationRequestParameters requestPar
             }
         }
 
-        private DateTimeOffset CalculateSuggestedCacheExpiry()
+        /// <summary>
+        /// Important note: we should not be suggesting expiration dates that are in the past, as it breaks some cache implementations.
+        /// </summary>
+        internal /* for testing */ static DateTimeOffset? CalculateSuggestedCacheExpiry(
+            ITokenCacheAccessor accessor, 
+            ICoreLogger logger)
         {
-            IEnumerable<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensWithNoLocks(true);
-            DateTimeOffset cacheExpiry = tokenCacheItems.Max(item => item.ExpiresOn);
-            return cacheExpiry;
+            // If we have refresh tokens in the cache, we cannot suggest expiration
+            // because refresh token expiration is not disclosed to SDKs and RTs are long lived anyway (3 months by default)
+            if (accessor.GetAllRefreshTokens().Count == 0)
+            {
+                IReadOnlyList<MsalAccessTokenCacheItem> tokenCacheItems = accessor.GetAllAccessTokens(optionalPartitionKey: null);
+                if (tokenCacheItems.Count == 0)
```

Comment:
I guess there might be multiple thread access issues here too (although probably unlikely)? Thread 1 returns 0 RTs, Thread 2 clears the cache and inserts RTs and ATs, Thread 1 operates on a different set of ATs than before step 1?

Reply:
Yes, this could explain why this issue occurs so rarely.